### PR TITLE
Admin secret implementation + test

### LINKED ISSUE #250: User impersonation
Is there a mecanism to impersonate/"sign in as a user" for support purposes in HBP?

What would it take to do this?

### PR DESCRIPTION
**Pull request by SkinyMonkey**
Close #250

This PR is not finished : i need to fix the test.
In the test I set a header and do a post request to HBP, but it does not seem to receive it for some reason ...
I'm not sure why ...
Help would be appreciated.

Also, the Commit log should be reworded as this commit is meants to squashed/fixed/amend

### CODE REVIEW

**Comment by SkinyMonkey** [comment]
Close #250

File: src/routes/auth/login.ts [hunk 1]
```diff
@@ -76,7 +80,7 @@ async function loginAccount({ body }: Request, res: Response): Promise<unknown>
     throw Boom.badRequest('Account is not activated.')
   }
 
-  if (!(await bcrypt.compare(password, password_hash))) {
+  if (!(await bcrypt.compare(password, password_hash)) && req.header(AdminSecretHeader) != ADMIN_SECRET) {
```
Comment (elitan) [inline src/routes/auth/login.ts:83]: should probably be `||` instead of `&&` here.

File: src/routes/auth/login.ts [hunk 2]
```diff
@@ -17,7 +17,11 @@ interface HasuraData {
   }
 }
 
-async function loginAccount({ body }: Request, res: Response): Promise<unknown> {
+export const AdminSecretHeader = 'X-HBP-Admin-Secret'
```
Comment (elitan) [inline src/routes/auth/login.ts:20]: How about naming the header to `x-hbp-admin-secret` in lower cases?

Comment (elitan) [reply src/routes/auth/login.ts:20]: And maybe rename this variable from `AdminSecretHeader` to something more descriptive like:

`NameOfAdminSecretHeader` (just an example)

The variable name `AdminSecretHeader` feels like we are setting both the key and value of the header.

File: src/routes/auth/auth.test.ts [hunk 3]
```diff
@@ -116,20 +113,30 @@ it('should complain about incorrect email', async () => {
...
 
-it('should decode a valid custom user claim', async () => {
-  const decodedJwt = JWT.decode(jwtToken) as Token
-  expect(decodedJwt[JWT_CLAIMS_NAMESPACE]).toBeObject()
-  // Test if the custom claims work
-  expect(decodedJwt[JWT_CLAIMS_NAMESPACE]['x-hasura-name']).toEqual('Test name')
+it('should sign the user in using the admin secret', async () => {
+  const { body, status } = await request.post('/auth/login')
+    .send({ email, password: '' })
+    .set(AdminSecretHeader, ADMIN_SECRET)
```
Comment (SkinyMonkey) [inline src/routes/auth/auth.test.ts:134]: This .set is not working, i don't know why

Comment (elitan) [reply src/routes/auth/auth.test.ts]: That's strange. The syntax looks OK.

Comment (elitan) [reply src/routes/auth/auth.test.ts]: Oh, apparently the order matters. Look here: https://github.com/visionmedia/supertest/issues/398#issuecomment-366403045

Comment (elitan) [reply src/routes/auth/auth.test.ts]: Wait, you got the correct order. :/

Comment (SkinyMonkey) [reply src/routes/auth/auth.test.ts]: I tried both without success :/

File: src/routes/auth/login.ts (hunk 1 above)
Comment (SkinyMonkey) [reply src/routes/auth/login.ts]: At first i thought you might be right but the tests broke after this change.
And it make sense:

The && in that case stands for:
"If neither the password and the header is right, send back the error"
A || would stands for :
"If the password or the header is not right, send back the error"
Which will not work if a client does not send the AdminSecret header

File: src/routes/auth/auth.test.ts (hunk 3 above)
Comment (AndonMitev) [reply src/routes/auth/auth.test.ts]: I still have this issue

**Comment by elitan** [comment]
Closed in favor of https://github.com/nhost/hasura-backend-plus/pull/421

**Pull request closed without merging**